### PR TITLE
fix(build): make rpath relative to ORIGIN for binaries

### PR DESCRIPTION
**Pull request by arezaii**
This PR seeks to make the build and installation of `chpl` and `chpldoc` be 
relocatable by modifying their `RPATH`s to be relative instead of absolute. 

Changes include:
* warn the user about an impending update to minimum `cmake` version `3.20`

* change the default installation mode to nothing

* generate a `BUILD_VERSION` file in `gen_release` when `developer` option
  specified as argument to `gen_release`

* `cmake` versions prior to `3.16` will still save absolute paths,
  and will fail to install using `--chpl-home` mode instead of `--prefix` mode

* change the installation location of the compiler library to be 
  `lib/chapel/major.minor/compiler` for `--prefix` installations

* for other installation types, the location has been updated to 
  `lib/compiler/host-bin-subdir` where `host-bin-subdir` is something
  like `linux64-x86_64`

* conditionally set relative `RPATH` to `$ORIGIN` or `@executable_path`,
  depending on if OS is linux or macos, respectively, and add a message
  when changing the RPATH for macos

* the driver of these changes is to fix Cray module builds for Chapel, which
  are built into RPMs and relocated

* fixed some line-lengths 

Linux Testing:

- [x] prefix build and install have absolute `RPATH` 
- [x] home build and install have relative `RPATH` starting with
- [x] when installation mode not set, build has relative `RPATH` starting with

Mac Testing:

- [x] prefix build and install have absolute `RPATH`
- [x] home build and install have relative `RPATH` starting with `@executable_path`
- [x] when installation mode not set, build has relative `RPATH` starting with `@executable_path`

reviewed by @mppf - thanks!

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -352,15 +352,18 @@ add_subdirectory(frontend)
...
-set_target_properties(chpl PROPERTIES
-                      INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
-set_target_properties(chpldoc PROPERTIES
-                      INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
-
 if (INSTALLATION_MODE STREQUAL "home")
+  set_target_properties(chpl PROPERTIES
+                        INSTALL_RPATH "$ORIGIN/../../lib")
+  set_target_properties(chpldoc PROPERTIES
+                        INSTALL_RPATH "$ORIGIN/../../lib")
```
[the remark this answers is not in the document]
Comment (mppf) [reply CMakeLists.txt:359]: Also `ORIGIN` is linux-specific and we'd need something else for Mac OS X.

See https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling .

My recommendation (to get cmake itself to do the platform-dependent parts):
 * if it is not configured for a prefix install, use [CMAKE_BUILD_RPATH_USE_ORIGIN](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_RPATH_USE_ORIGIN.html#variable:CMAKE_BUILD_RPATH_USE_ORIGIN) and instruct cmake not to change the rpath when installing
 * if it is configured for a prefix install, use the current absolute path approach, where it is included at build time and then modified at install time.

[the remark this answers is not in the document]
Comment (mppf) [reply CMakeLists.txt:359]: It looks like `liblibdyno.so` is in `lib/compiler/linux64-x86_64/liblibdyno.so` at build time. After a `--chpl-home` installation, it is installed there *but also copied* to `lib/` in the installation directory. I think this should not change between being installed vs not. The `linux64-x86_64` can remain in the path and the relative part should just be at the beginning.

File: util/buildRelease/gen_release [hunk 2]
```diff
@@ -100,6 +100,7 @@ if (exists($ENV{"CHPL_GEN_RELEASE_NO_CLONE"})) {
 
     if ($version eq "" or $version eq "developer") {
       print "[gen_release] Copying BUILD_VERSION file.\n";
+      SystemOrDie("util/config/write-git-sha frontend/lib/util --build-version --chpl-home=$archive_dir");
```
Comment (mppf) [inline util/buildRelease/gen_release:103]: Worth a note in the PR message mentioning this change

**Review by mppf — APPROVED** [review]
Looks good, thanks!

**Comment by arezaii** [comment]
noting that you can view the RPATH like so:

Linux:
```
objdump -x /path/to/exe | grep PATH
```
MacOS:
```
otool -l /path/to/exe | grep path
```

**Pull request merged**